src/services/EthereumService.ts

```diff
@@ -213,238 +341,31 @@ export class EthereumService {
-            },
-          },
-        });
-
-        if (wasAdded) {
-          this.setMetamaskHasToken(tokenAddress);
-        }
+        const provider = detectEthereumProvider ? (await detectEthereumProvider({ mustBeMetaMask: true })) as any : undefined;
+        this.metaMaskWalletProvider = provider;
       } catch (error) {
-        console.log(error);
+        this.consoleLogService.logObject(error.message, error, "error");
       }
     }
+  }
 
-    return wasAdded;
+  private async addWalletProviderListeners(): Promise<void> {
+    await this.removeWalletProviderListeners();
+
+    await this.ensureMetaMaskWalletProvider();
+    this.metaMaskWalletProvider.on("accountsChanged", this.handleAccountsChanged);
+    this.metaMaskWalletProvider.on("chainChanged", this.handleChainChanged);
+    this.metaMaskWalletProvider.on("disconnect", this.handleDisconnect);
   }
 
-  public getMetamaskHasToken(tokenAddress: Address): boolean {
-    if (!this.defaultAccountAddress) {
-      throw new Error("metamaskHasToken: no account");
-    }
-    return !!this.storageService.lsGet(this.getKeyForMetamaskHasToken(tokenAddress));
-  }
-
-  /**
-   * so unit tests will be able to complete
-   */
-  public dispose(): void {
-    this.readOnlyProvider.off("block", (blockNumber: number) => this.handleNewBlock(blockNumber));
-  }
-
-  public async getBlock(blockNumber: number): Promise<IBlockInfo> {
-    const block = await this.readOnlyProvider.getBlock(blockNumber) as unknown as IBlockInfo;
-    block.blockDate = new Date(block.timestamp * 1000);
-    return block;
-  }
-
-  public getEtherscanLink(addressOrHash: Address | Hash, tx = false): string {
-    let targetedNetwork = EthereumService.targetedNetwork as string;
-    if (targetedNetwork === Networks.Mainnet) {
-      targetedNetwork = "";
-    } else {
-      targetedNetwork = targetedNetwork + ".";
-    }
-
-    return `http://${targetedNetwork}etherscan.io/${tx ? "tx" : "address"}/${addressOrHash}`;
-  }
-
-  /**
-   * returns ENS if the address maps to one
-   * @param address
-   * @returns null if there is no ENS
-   */
-  public getEnsForAddress(address: Address): Promise<string> {
-    return this.readOnlyProvider?.lookupAddress(address)
-      .catch(() => null);
-  }
-
-  /**
-   * Returns address that is represented by the ENS.
-   * Returns null if it can't resolve the ENS to an address
-   * Returns address if it already is an address
-   */
-  public getAddressForEns(ens: string): Promise<Address> {
-    /**
-     * returns the address if ens already is an address
-     */
-    return this.readOnlyProvider?.resolveName(ens)
-      .catch(() => null); // is neither address nor ENS
-  }
-
-  private handleNewBlock = async (blockNumber: number): Promise<void> => {
-    const block = await this.getBlock(blockNumber);
-    this.lastBlock = block;
-    this.eventAggregator.publish("Network.NewBlock", block);
-  };
-
-  // private cachedProviderKey = "cachedWalletProvider";
```

Comment:
https://github.com/PrimeDAO/prime-deals-dapp/pull/1195/files/2ce2449d45a041d9052b4adc4cbbc67bf8b6b49a#r954892172

@giladt that code is from here